src/components/createFormattedComponent.tsx

```diff
@@ -1,46 +1,49 @@
 import * as React from 'react';
+import {invariantIntlContext} from '../utils';
 import {IntlShape, FormatDateOptions, FormatNumberOptions} from '../types';
-import withIntl, {WrappedComponentProps} from './injectIntl';
+import {Context} from './injectIntl';
 
-export default function createFormattedComponent<
-  T extends 'formatDate' | 'formatTime' | 'formatNumber'
->(type: T) {
-  type FormatOptions<T> = T extends 'formatDate'
-    ? FormatDateOptions
-    : T extends 'formatTime'
-    ? FormatDateOptions
-    : FormatNumberOptions;
-  type FormatFn = IntlShape[T];
-  type Props = FormatOptions<T> &
-    WrappedComponentProps & {
-      value: Parameters<FormatFn>[0];
-      children?(val: string): React.ReactElement | null;
-    };
-  const Component: React.FC<Props> = props => {
-    const {
-      value,
-      children,
-      intl: {[type]: formatFn, textComponent: Text},
-    } = props;
+const displayNames: {
+  formatDate: 'FormattedDate';
+  formatTime: 'FormattedTime';
+  formatNumber: 'FormattedNumber';
+} = {
+  formatDate: 'FormattedDate',
+  formatTime: 'FormattedTime',
+  formatNumber: 'FormattedNumber',
+};
 
-    let formattedValue = formatFn(value as any, props);
+type Formatter = {
+  formatDate: FormatDateOptions;
+  formatTime: FormatDateOptions;
+  formatNumber: FormatNumberOptions;
+};
 
-    if (typeof children === 'function') {
-      return children(formattedValue);
-    }
-    if (Text) {
-      return <Text>{formattedValue}</Text>;
-    }
-    // Work around @types/react where React.FC cannot return string
-    return formattedValue as any;
-  };
-  Component.displayName =
-    type === 'formatDate'
-      ? 'FormattedDate'
-      : type === 'formatTime'
-      ? 'FormattedTime'
-      : 'FormattedNumber';
-  return {
-    Component: withIntl(Component),
+export default function createFormattedComponent<Name extends keyof Formatter>(
+  name: Name
+) {
+  type Options = Formatter[Name];
+  type FormatFn = IntlShape[Name];
+  type Props = Options & {
+    value: Parameters<FormatFn>[0];
+    children?: (val: string) => React.ReactElement | null;
   };
+
+  const Component = (props: Props) => (
```

Comment:
#1416 not use any then
```typescript
const Component: React.FC<Props> = (props) => {
   // ... code ...
}
```

`React.FC` definition: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/981cc6d51eb213943a3606c14ad75d65ea6b360f/types/react/index.d.ts#L493-L501

Reply:
Hmm what issue should this solve exactly? Also, can you explain how using `React.FC` is different from `props: Props`? 🤔

Reply:
expand `React.FC<Props>` definition got `(props: Props, context?: any) => React.ReactElement | null`

Reply:
yeah `const Component: React.FC<Props> = props =>` gives u the correct type definition

Reply:
Just to clarify, what do you mean by `the correct type definition`?
Using `props: Props` is not wrong, this is also what you [find in the TS docs](https://www.typescriptlang.org/docs/handbook/jsx.html#function-component)

<img width="451" alt="image" src="https://user-images.githubusercontent.com/1110551/62937683-a246b900-bdcd-11e9-8be8-5ec09472faf0.png">

I'm also not against using `React.FC` if you want to, I just want to understand what are the benefits of using that compared to `props: Props`.

Reply:
https://github.com/formatjs/react-intl/pull/1415/commits/11335f78c3c04b91078ca311650a5dcfd1069a03